### DISCUSSION TITLE
accessible page numbers / page breaks

### DISCUSSION THREAD

**Discussion by danielweck** [edrlab/thorium-reader#1799] [Ideas]
From issue https://github.com/edrlab/thorium-reader/issues/1662

I don't think the page number issue is usably resolved with Thorium 2,0, and so the issue should be reopened.  I think I found an obvious solution.

First, I've been discussing this Thorium and page number issue with a blind researcher in a university. This is the message being relayed to this discussion:

“For students, researchers and scholars with print disabilities, every accessibility barrier we encounter slows us down and takes time away from the most valuable tasks we should be engaged in. Having a robust feature that provides us with page numbers would not only enable us to keep up with the referencing conventions of our disciplines, but would help to make citation tasks faster and more efficient. In addition, having access to page numbers quickly and easily lets us check the citations used by others, so we can focus our energies on keeping up with scholarly sources in our fields.”

It occurred to me, maybe there is an obvious solution. Even if page numbers are in the EPUB file they are not being rendered in the content.  Here's an example from an academic EPUB:

`span.pagebreak-rw { width: 0; font-size: 0; line-height: 0; height: 0; visibility: hidden; float: right; }
`

Could Thorium have a couple of settings?

1. Show page number with message, such as "Start page"
2. Show page numbers with line breaks

The image below shows a page from an EPUB that has page break. It starts with: [Start Page 15]. I have verified this is correct with the physical item.

![image](https://user-images.githubusercontent.com/8799391/180653074-2a45d566-fb28-4f4a-9a70-57157007fbeb.png)

This second image shows some text from an EPUB and in the middle of a sentence is: [Start page 16].  I have verified this is correct with the physical item.

![image](https://user-images.githubusercontent.com/8799391/180653658-0f479d0c-2157-4c1f-9158-0e8f375bdc2b.png)

This third image shows some text from an EPUB. In the middle of a sentence "21" appears with paragraph break before it and after it for sighted scholars.

![image](https://user-images.githubusercontent.com/8799391/180654523-46f09730-be1b-4159-9efb-53ab30b4a098.png)

Button 1 would address the researcher's comment.

Button 2 would make Thorium more usable friendly for sighted researchers in universities who also need page numbers for quotation purposes. 

If a reader doesn't want page numbers, then keep both buttons off.

**Comment by spudthebud** [comment]
I'm restarting this discussion.  I have spoken with multiple blind academics, from undergraduate student to masters student to doctoral student to faculty, who have expressed great frustration at the great difficulty and often impossibility of completing the task of getting page numbers from academic reading materials. This includes, but goes beyond text books, and includes academic eBooks in EPUB form and journal articles in EPUB form.

**Reply by gautierchomel** [reply]
There are two different cases.

1. the publisher inserted page numbers in reference to printed version. In this case Thorium reproduce it in the [WhereAmI](https://edrlab.github.io/thorium-reader-doc/210_reading/212_whereAmI/) function and allows page navigation thru [Go to page menu](https://edrlab.github.io/thorium-reader-doc/210_reading/211_moving#the-reference-pages). A button to render pagination in the content may be a plus.
2. the publisher did not inserted page numbers in the publication and Thorium can not do anything. In such a case reclamation must be done to book seller and publisher.

The page number inclusion information is currently not displayed by vendors but will soon be displayed in Thorium book information panel. However here too we can only display an information present in the file, so Publishers must inform using the [schema.org metadata accessibility feature](https://schema.org/accessibilityFeature).

**Comment by spudthebud** [comment]
1. The WhereAmI function is not practically usable with a screen reader as the screen reader's focus is lost after moving it. Also, it won't produce correct results consistently because Thorium cannot know where the screen reader's focus is within the viewport. see #1662 .  This is why the option to let the reader toggle rendering page number in text is raised here. 

2. Agreed. If the publisher didn't add page numbers, not much can be done.  But acquisition departments and purchasers can notifying them of the problem.

**Reply by danielweck** [reply]
> The WhereAmI function is not practically usable with a screen reader as the screen reader's focus is lost after moving it.

I assume you mean the `SHIFT CONTROL i` keyboard shortcut. Have you tried `SHIFT CONTROL K` which displays an overlay announcement which does not change keyboard focus?

**Reply by danielweck** [reply]
> Also, it won't produce correct results consistently because Thorium cannot know where the screen reader's focus is within the viewport.

Just for new readers / visitors in this discussion: this is broadly-speaking a correct statement, but if screen reader users can click on text just like mouse / pointing device users, then Thorium detects the click event and infers the reading position from it (i.e. the HTML element that contains the clicked text becomes the current reading location, which can be used for resuming and bookmarking)

**Reply by danielweck** [reply]
> This is why the option to let the reader toggle rendering page number in text is raised here.

Thank you for framing the feature request once again. Let's see what we can do to force the accessible display (i.e. visual presentation and screen reader labelling) of interspersed page breaks in HTML documents.

**Comment by spudthebud** [comment]
As I indicated in 1662, Shift + Control + K isn't reliable.  I spoke with another blind academic (PhD) today who added more voiced  frustration around getting page numbers in university.  Why don't we go for the gold medal with this feature?  And can we name the feature "Maple Leaf" in the development phase in recognition of the Canadian blind students and scholars who have spoken out?

**Reply by gautierchomel** [reply]
Thanks, 

I would like to understand why Shift + Control + K isn't reliable. In #1662 you mention that _page is announced and cursor is kept in the same spot_. You also mention that _page number announced does not match the printed book_, this point is up to the publisher or producer of the book. Is there another point making Shift + Control + K not usable ? 

I'm not opposed to having a display option for page numbers, it exposes a hidden element and that's always useful, including for pedagogical aspects. 

My point is that we still have to make sure one can know where he is without having page number displayed or enunciated at each page. That why i would like to understand what's wrong with Shift + Control + K (and think we have to focus on fixing it).

For context, before 2015 I produced adapted textbooks in epubs for blind students, I included page numbers in the content because no reading system was displaying them. It was better than nothing but the reading experience is not good and once a user knows how to call page number when needed he will never display it again. As page number may be in the middle of a word you'll get very strange cases and even if you manage to put page number at logical break points (like end of sentence), it's a lost of concentration to get "page xx" enunciated at every page.

**Comment by spudthebud** [comment]
Attached is a video showing the problem: the get page results aren't accurate. I try to provide audio description for non-visual access. 

https://user-images.githubusercontent.com/8799391/189116783-38cd2702-7896-40f0-bc5a-40249e76552c.mp4

I described this problem in #1662 and @danielweck responded [explaining the technical problem](https://github.com/edrlab/thorium-reader/issues/1662#issuecomment-1161534338) "Thorium estimates the reading location by looking at the topmost text in the current visible viewport". 

This prompted my suggestion of giving the user the option to toggle on or off the page number in text.

**Reply by gautierchomel** [reply]
Many thanks, 
I do reproduce using NVDA on windows 10, page number announced thru Shift + Control + K is not correct. The page break code is `<span epub:type="pagebreak" role="doc-pagebreak" id="page_2" class="page" title="2"/>` and NVDA says "2 separateur" (separator 2) but _Where am I_  gives a false number. This is tricky! 

I'm in favor of an option to toggle on or off the page number in text, but if _where am I_ does not work correctly we have to inform about it.

**Reply by gautierchomel** [reply]
Added to documentation : https://github.com/edrlab/thorium-reader-doc/pull/13/commits/ab3169602cd43f94f1f3870996a3ef69a2532480

**Comment by danielweck** [comment]
Here is a related discussion about title attribute, ARIA label vs. textual content in page break spans:
https://github.com/daisy/ace/issues/355

**Comment by danielweck** [comment]
[spudthebud](https://github.com/spudthebud) what I am failing to clearly understand is why would the reading system need to force the visual rendering of page breaks in HTML documents, when normally a screen reader would be expected to pick-up the ARIA label or title attribute of `span` elements (for example) that do not have explicit text content / have an empty child.

**Comment by spudthebud** [comment]
Would aria labels or titles be searchable by a screen reader's own built-in search feature, which is very popular? 

And not everyone who has a print disability and needs to get the page number is blind.   For example, people with low vision who are reading with magnification and reflow might not have a text-to-speech synthesizer running.

I don't think this suggestion is radical or new.  The Daisy Consortium's WordToEPUB tool supports displaying the page numbers. It allows you to specify the text that comes before and after the page number.

And some platforms provided by academic journal publishers do it.

For example, I believe this article "[Rest in Pieces: Body Donation in Mid-Twentieth Century America"]( by https://muse.jhu.edu/article/861408) is not behind a subscription pay-wall.  If so, the image below shows it

<img width="861" alt="description follows" src="https://user-images.githubusercontent.com/8799391/189955464-5f778a37-6541-492e-9b37-dc004a605a9f.png">

The image shows a paragraph of text from an HTML version of an academic journal article. The page number is added as text in the visual interface.

**Reply by danielweck** [reply]
The mechanism exists for EPUB authors to display page breaks, if they want to. But if instead content creators decide to visually hide page breaks (typically, when they are inlined in between words in the middle of sentences), then only screen readers can access them (or more generally speaking, AT), via ARIA label. Because such empty span HTML element can be linked from the EPUB pagelist, a title attribute is often provided too (if I remember correctly).
So the question here is (please tell me if I am misguided): should the reading system allow forcing the visual presentation of hidden page breaks (assuming this is technically possible, in a reliable manner), even though the "where am I" problem seems to affect only screen reader users due to not being able to point / select text in the currently displayed page spread or scrolling viewport extent?

**Comment by spudthebud** [comment]
If I understand your question correctly: The reading system should allow the option for readers to display page numbers within the text.  I am not commenting on page breaks.

The reason is: People can have sight and disabilities that make it difficult or impossible to use the mouse.

This 11 minute documentary "[We Regret to Inform You...](https://www.nfb.ca/film/we_regret_to_inform_you/)" is by Dr. Heidi Janz, who is a Canadian disabilities scholar and playwright. She documents some of the experiences of being a disabled academic in Canada. The video has an audio description button. At about 7 minutes 40 seconds we see how Dr. Janz types. It reminds us that not everyone uses input devices in ways we might expect.    I would not assume that every sighted disabled person can use a mouse. There is a WCAG 2.0 success criteria provides a reminder to that effect too.

**Comment by gautierchomel** [comment]
Trying to resume (please correct if needed): 

## use case one : AT users 

- **State**:  AT users have Screen readers announcing the **title** value of a **page break** when he meets it. 
- **Problem**: The user who want to know at which page he is standing has to go back and forward to find this information because we can't actually know where is the screen reader.  
- **Solution (partial)**: This limitation is now documented in the  [WhereAmI section](https://edrlab.github.io/thorium-reader-doc/210_reading/212_whereAmI/). 
- **Possible improvement**: _We can open the discussion with screen readers providers and ask them to implement WhereAmI or provide information to the Reading System so we can accurately determine the reading position._

_Note: screen reader covers non visual reading and interactions including eyestracking and others input methods._  

## use case two : non AT users  

- **State**: Non AT users can use [WhereAmI ](https://edrlab.github.io/thorium-reader-doc/210_reading/212_whereAmI/) to acknowledge their position, this is documented and can be learned. 
- **Problem**: because action is required, the reading experience can be untimently interrupted to find this information. 
- **Possible improvement**: May an improvement be a thick box option _Display the value of pageBreak in the margin_? This way it does not break content (exemple : _This phrase I am reading has page numbers breaking a wo**page 2**rd. This one has a page number breaking the **page 3** sentence_) and a non AT user can quickly identify at what page he is standing.

**Comment by spudthebud** [comment]
I'll get back to you on this as soon as possible

**Comment by spudthebud** [comment]
**Use case  one and two:  AT and Non AT Users**
The WhereAmI feature will not provide the correct page number in some cases, as documented in the video on Sept 8th.

If the user cannot get the correct page number, then there is a very serious problem, especially in academic environments. 

A mechanism that allows the user to toggle on or off the display of page number within the visual interface (and aural interface), would solve the problem.  On the Sept 13 comment, I document how a major academic platform adds the page number into the interface solving the problem. 

If the user can toggle the display of page numbers on or off in the text, then they decide if they want a text-display that supports reading comprehension (page numbers off) or that supports citation (page numbers on).

Again, I quote a blind user:

_“For students, researchers and scholars with print disabilities, every accessibility barrier we encounter slows us down and takes time away from the most valuable tasks we should be engaged in. Having a robust feature that provides us with page numbers would not only enable us to keep up with the referencing conventions of our disciplines, but would help to make citation tasks faster and more efficient. In addition, having access to page numbers quickly and easily lets us check the citations used by others, so we can focus our energies on keeping up with scholarly sources in our fields.”_

This is not the only blind scholar who has told me getting page numbers is a problem.

And here's a Q and A from a library conference presentation. It's a blind and sighted team presenting on problems in university-level library reading materials, who are responding to problems presented to the blind member:   https://youtu.be/0uC-arfXCmQ?t=1092

**Comment by gautierchomel** [comment]
I may miss again something but I don't see how this feature would solve the problem:

* Where Am I is not accurate only with screen readers (i.e. the problem does not exist for non AT users).
* Screen readers already announce the page break (i.e. the feature exist into AT).

To solve the problem, the solution is to get Screen readers to expose where they are.

An improvement could be to add page numbers in the margins for a quick visual reference.

**Comment by spudthebud** [comment]
Attached is a video of an example. It includes audio description.

It shows how adding a feature that toggles page number into the text would make it easier for blind screen readers users and sighted print disabled users to complete the tasks of getting page numbers for citation and also the task of navigating by page numbers.  

The technique recognizes that people read for comprehension and also read for citation.

This technique requires content authors to include page number information in their EPUB files.

The proposal of advocating and waiting for developers at multiple screen reader companies or groups (e.g., Vispero, Apple, Microsoft, NVDA, etc.) to expose the virtual cursor will take a very long time. During which time, many university students, staff, and faculty will face significant disadvantages.

https://user-images.githubusercontent.com/8799391/199480880-3528001f-fad8-4d77-836a-b9819addd50e.mp4

**Comment by gautierchomel** [comment]
Thanks for the video. 

* one can actually search a page by using the [page navigation](https://edrlab.github.io/thorium-reader-doc/210_reading/211_moving/#the-reference-pages) panel (shortcut Ctrl+Shift+P). 
* Your example shows page numbers well placed. I have produced a lot of epubs with visual page numbers and I had to move some page numbers to make sure they don't cut a word otherwise I got a _wo[page 108]rd_ occurrences. I can do this as a creator but how could we address that from the reading system?
* What do you think about the margin display option? [An exemple can be seen on Cairn International](https://www.cairn-int.info/journal-agora-debats-jeunesses-2020-3-page-143.htm#pa15) although they reference paragraph number instead of page number.

**Comment by spudthebud** [comment]
I have a meeting with a blind researcher.  If there's time, I may be able to bring this up.

**Comment by spudthebud** [comment]
1.  The video (with my audio description for blind viewers) below shows that Control + Shift + P does not work with JAWS and it does not work in  a usable way for sighted printed disabled readers as there is no visual indicator of where the page actually starts. 
2.  I spoke with two academic publishers, one in Canada and one in Australia, and asked about page breaks that split words. Both said that isn't common, at least in academic publishing.  
3.  Margin display option isn't accessible: Magnification loses the page number off screen.  However, magnification would work fine with the technique I suggested.  


https://user-images.githubusercontent.com/8799391/200445061-ee2c97ba-81b7-496a-a48c-5e19cec978a6.mp4

**Comment by gautierchomel** [comment]
Thank you, we have to check where is the page break into this document. Could you send me a copy of it at gautier.chomel @ edrlab.org? Alternatively you can find this by opening the EPUB file with Sigil and search for arguments `"role="doc-pagebreak"`.

**Comment by spudthebud** [comment]
I'm working on it.

**Comment by spudthebud** [comment]
Here's the code snippet from the EPUB.

`It irrevocably changed the way the great majority of the community saw <span class="pagebreak-rw" id="pagebreak-rw_19" epub:type="pagebreak" role="doc-break">19</span>the gun lobby. These men`

Also, here's a video from a recent panel presentation from a Canadian university library conference.  The speakers, other than the facilitator, are blind scholars, at all levels of their academic journey, from recent undergrad to professor. 

Dr. Cynthia Bruce, who is a professor from Concordia University, explains the problem of getting page numbers at the 12:10 mark in the [video](https://vimeo.com/770786119).  She explains "_Page numbers are another significant issue. And I know that seems small but it is actually monumental for us as we try to navigate these documents.... This threatens my capacity to cite sources accurately and it can be one of the most time consuming barriers to remediate_".

I believe giving the reader the option to toggle the presence or absence of page numbers in the document, such as **Start page 283**, when available in the EPUB, would address Dr. Bruce's concern, as well as the concern of the researcher at the start of this issue thread.

**Reply by gautierchomel** [reply]
Thank you, there is an error in this document, `"role="doc-break"` should be `"role="doc-pagebreak"`. See dedicated [doc-pagebreak page on the Daisy knowledge base](http://kb.daisy.org/publishing/docs/html/dpub-aria/doc-pagebreak.html) for more info.

**Reply by spudthebud** [reply]
It was a typo when I transcribed it.  The role in the EPBUB is: role="doc-pagebreak"

**Comment by gautierchomel** [comment]
We had a discussion on the subject with @GeorgeKerscher. Georges mentioned that a way to get precise screenreader position is to simulate a mouse clic. I tested with NVDA and got good results. This is documented in the 
[Where Am I feature documentation of Thorium](https://edrlab.github.io/thorium-reader-doc/210_reading/212_whereAmI/#access-more-informations).

Short how to is as follow : _Press the NVDA + Enter keys to save the current position. Then Ctrl + Shift + I to place the cursor in the Progress section of the dialog box containing information about the information dialog box._

Of course this does not totally solve the problem but at least allow screen reader users to know their exact page position. Feedback about how it works with Jaws and VoiceOver is very welcome.

We are currently looking for funding to add annotations features to Thorium. Page position will be part of the work.

Still forcing inclusion of a page number in the content is not a sufficient solution as it will break words in some books. I prefer adding page numbers in the margins and make them links to the exact page position in the content.

**Reply by spudthebud** [reply]
I'll try the suggestion with JAWS. 

As a reminder, I checked with two academic publishers, one in Canada and one in Australia, and both said page breaks splitting words is not very common.  Giving people the option about how page numbers are displayed or not is giving the reader choice about their reading experience.

**Reply by spudthebud** [reply]
I think George was suggesting routing the JAWS cursor to the virtual PC cursor and then use JAWS + numpad minus to left click.    I  tried left mouse click on the word before a page break.

### Results

- Thorium could not get accurate page numbers using mouse click then Control + shift + i

### Conclusion
I have not reviewed the efficiency or effectiveness of Thorium's technique to get page numbers for citation tasks.  However, based on the steps I've done, I found that Thorium's technique could not accurately get page numbers. Therefore, I have to conclude the technique does not work.

Additionally, the current technique will not work for sighted print-disabled readers who don't have hand motor skills to use a mouse or track pad (I shared video example [earlier](https://github.com/edrlab/thorium-reader/discussions/1799#discussioncomment-3639807)) 

Techniques that disadvantage print-disabled readers are flawed and should not be relied on in reading systems.  Funders and developers need to be encouraged to identify and support techniques that remove these sorts of disadvantages.  I have proposed an [alternative technique](https://github.com/edrlab/thorium-reader/discussions/1799#discussioncomment-4036093) to getting page numbers. I have documented a real example of it in [another post](https://github.com/w3c/publishingcg/issues/51#issuecomment-1356353856).

**Comment by gautierchomel** [comment]
Ho, one thing to mention, Thorium detects if a screen reader is active at startup. As a consequence starting the screen reader once Thorium is launched will cause synchronization problems between both. 

This may be what happens in the demo video. Could you check with screen reader active before launching Thorium? 

I do not reproduce such errors with test files I have.

**Comment by gautierchomel** [comment]
A similar discussion is taking place at W3C publishing community group. 

A readium based [experimental implementation](https://github.com/w3c/publishingcg/issues/51#issuecomment-1355358568) can be seen there.

**Reply by spudthebud** [reply]
Thanks for the reference. Much appreciated.

For the record, here's [what I posted](https://github.com/w3c/publishingcg/issues/51#issuecomment-1356353856) on W3C Publishing Community Group thread.

**Comment by danielweck** [comment]
From https://github.com/edrlab/thorium-reader/issues/2567
@ddfreiling :

A highly requested feature from our study-book users is being able to visually see the page numbers of the book.

Could this be made as a togglable feature? We'd love to give the implementation a shot, if EDRLab is onboard with having this feature in Thorium Reader.

**Comment by Lewiscowles1986** [comment]
A Few questions on this:

1. Does Thorium obey `page-break-after: always;` CSS?
2. Would Thorium obey `<br clear="all">`?

If either of these were true, then Authors (sadly I think this might be an authorship problem); could augment their books to have the page numbers, which would add enough artifacts to be able to detect either of the above patterns, and infer page numbers from it.

Last question. Could anchors be used, with a table-of-contents or index, that is a nav element, with role="doc-toc" as pointed out in https://kb.daisy.org/publishing/docs/html/dpub-aria/doc-toc.html this combined with explicit id's to anchor to should allow fragment based navigation and page based navigation. (this is more about getting to a page)
https://www.w3.org/TR/dpub-aria-1.0/#doc-toc also covers as a standards element.

**Reply by danielweck** [reply]
> Does Thorium obey `page-break-after: always;` CSS?

Hello, Thorium itself doesn't specifically handle this styling construct, the Chromium web browser engine does. Thorium does inject CSS column logic to emulate visual / "screen" pagination (i.e. not actual pagination in the EPUB3 sense, which mirrors print page breaks). Thorium also injects ReadiumCSS so this can potentially interfere with authored styles https://github.com/readium/css

> Would Thorium obey `<br clear="all">`?

Once again this needs to be practically tested to observe the real-world outcome. I assume this might have an impact not only in "paginated" mode (CSS columns) but also in scroll mode (vertically flowing HTML document).

> this combined with explicit id's to anchor to should allow fragment based navigation and page based navigation

In EPUB3, the page breaks and page list constructs are all "fragment"-based (i.e. `id` attributes on HTML elements). Thorium supports this functionality already, could you please explain what seems to be missing?
